### PR TITLE
Don't warn for an unused type alias if its member type is used

### PR DESCRIPTION
**Pull request by jxnu-liguobin**
scala/bug#12600

### CODE REVIEW

**Comment by som-snytt** [comment]
A warning was lost in `neg/warn-unused-privates`. I don't know the correct answer without research.

**Comment by SethTisue** [comment]
@som-snytt interested in having a look at this?

**Comment by som-snytt** [comment]
I will make an attempt at understanding the LOC.

**Comment by SethTisue** [comment]
Ping. (But if you're unsure or have too much else on your plate, don't hesitate to decline.)

**Comment by som-snytt** [comment]
I should have tried it today while I was recovering from days of high fever and still incoherent and unable to understand anything. But it's not 6 months yet, we'll definitely make it into the milestone.

**Comment by som-snytt** [comment]
I meant the next milestone.

**Comment by SethTisue** [comment]
@som-snytt ping

File: test/files/pos/t12600.scala [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// scalac: -Wunused:privates,locals,patvars
```
Comment (som-snytt) [inline test/files/pos/t12600.scala:1]: The pos test does not check output, only success/fail, so `-Werror` is necessary to fail on warning.

File: src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala [hunk 2]
```diff
@@ -600,7 +605,9 @@ trait TypeDiagnostics extends splain.SplainDiagnostics {
         && !isSuppressed(m)
         && !m.isTypeParameterOrSkolem // would be nice to improve this
         && (m.isPrivate || m.isLocalToBlock || isEffectivelyPrivate(m))
-        && !(treeTypes.exists(_.exists(_.typeSymbolDirect == m)))
+        && {
+        !treeTypes.exists(_.exists(_.typeSymbolDirect == m))
+      }
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala:610]: spurious change, could just remove parens.

File: src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala [hunk 3]
```diff
@@ -592,6 +592,11 @@ trait TypeDiagnostics extends splain.SplainDiagnostics {
           case _                  => ()
         }
       }
+      if (sym != null && sym.tpe.typeSymbol.isRefinementClass) {
+        // see scala/bug#12600, We should emit warn for alias which is a refined type
+        // but because of type discard, we can not check if it was used in type parameters, we also cannot exactly check if there is a duplicate alias.
+        treeTypes += sym.tpe
+      }
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala:599]: This outcome would be unsatisfactory because now dotty does it.

**Review by som-snytt — COMMENTED** [review]
Sorry for the delayed review. It was because I would need several hours to understand how type trees work. I've proposed a fix, so we can close this non-fix.

**Comment by som-snytt** [comment]
superseded by https://github.com/scala/scala/pull/10494

**Pull request closed without merging**